src/txreconciliation.h

```diff
@@ -0,0 +1,585 @@
+     */
+    std::chrono::microseconds m_next_recon_respond{0};
+    std::chrono::microseconds NextReconRespond()
+    {
+        auto current_time = GetTime<std::chrono::microseconds>();
+        if (m_next_recon_respond < current_time) {
+            m_next_recon_respond = current_time + RECON_RESPONSE_INTERVAL;
+        }
+        return m_next_recon_respond;
+    }
+
+    public:
+
+    TxReconciliationTracker() {};
+
+    /**
+     * TODO: document
+     */
+    std::tuple<bool, bool, uint32_t, uint64_t> SuggestReconciling(const NodeId peer_id, bool inbound);
+
+    /**
+     * If a peer was previously initiated for reconciliations, get its current reconciliation state.
+     * Note that the returned instance is read-only and modifying it won't alter the actual state.
+     */
+    bool EnableReconciliationSupport(const NodeId peer_id, bool inbound,
+        bool recon_requestor, bool recon_responder, uint32_t recon_version, uint64_t remote_salt,
+        size_t outbound_flooders);
+
+    /**
+     * If a it's time to request a reconciliation from the peer, this function will return the
+     * details of our local state, which should be communicated to the peer so that they better
+     * know what we need.
+     */
+    Optional<std::pair<uint16_t, uint16_t>> MaybeRequestReconciliation(const NodeId peer_id);
+
+    /**
+     * Record an (expected) reconciliation request with parameters to respond when time comes. All
+     * initial reconciliation responses will be done at the same time to prevent privacy leaks.
+     */
+    void HandleReconciliationRequest(const NodeId peer_id, uint16_t peer_recon_set_size, uint16_t peer_q);
+
+    /**
+     * TODO document
+     */
+    Optional<std::vector<uint8_t>> MaybeRespondToReconciliationRequest(const NodeId peer_id);
+
+    /**
+     * TODO document
+     */
+    std::vector<uint256> FinalizeIncomingReconciliation(const NodeId peer_id,
+        bool recon_result, const std::vector<uint32_t>& ask_shortids);
+
+    /**
+     * Received a response to the reconciliation request. May leak tx-related privacy if we announce
+     * local transactions right away, in case the peer is strategic about sending sketches to us via
+     * different connections (requires attacker to occupy multiple outgoing connections).
+     * Returns a response we should send to the peer, and the transactions we should announce.
+     */
+    Optional<std::tuple<bool, bool, std::vector<uint32_t>, std::vector<uint256>>> HandleSketch(
+        const NodeId peer_id, int common_version, std::vector<uint8_t>& skdata);
+
+    /**
+     * Peer requesting extension after initial reconciliation failed on their side.
+     * No privacy leak can happen here because sketch extension is constructed over the snapshot.
+     */
+    void HandleIncomingExtensionRequest(const NodeId peer_id);
+
+
+    Optional<ReconciliationState> GetPeerState(const NodeId peer_id) const
```

Comment:
This is piercing the encapsulation. The client should never handle a `ReconciliationState` object. There should be a class interface function in `TxReconciliationTracker` for the client to get the data that it needs.